File: R/TOSTtwo.R
#' TOST function for an independent t-test (Cohen's d)
#' @param m1 mean of group 1
#' @param m2 mean of group 2
#' @param sd1 standard deviation of group 1
#' @param sd2 standard deviation of group 2
#' @param n1 sample size in group 1
#' @param n2 sample size in group 2
#' @param low_eqbound_d lower equivalence bounds (e.g., -0.5) expressed in standardized mean difference (Cohen's d)
#' @param high_eqbound_d upper equivalence bounds (e.g., 0.5) expressed in standardized mean difference (Cohen's d)
#' @param alpha alpha level (default = 0.05)
#' @param var.equal logical variable indicating whether equal variances assumption is assumed to be TRUE or FALSE.  Defaults to FALSE.
#' @return Returns dataframe with t-value, degrees of freedom, p-value, and lower and upper limit of confidence interval
#' @examples
#' TOSTtwo(m1=7.83,m2=7.98,sd1=1.21,sd2=1.29,n1=400,n2=400,low_eqbound_d=-0.199897678576151, high_eqbound_d=0.199897678576151,alpha=0.05, var.equal=TRUE)
#' @export
#' 

TOSTtwo<-function(m1,m2,sd1,sd2,n1,n2,low_eqbound_d, high_eqbound_d, alpha, var.equal){
  if(missing(alpha)) {
    alpha<-0.05
  }
  if(missing(var.equal)) {
    var.equal<-FALSE
  }
  if(var.equal==TRUE) {
    sdpooled<-sqrt((((n1 - 1)*(sd1^2)) + (n2 - 1)*(sd2^2))/((n1+n2)-2)) #calculate sd pooled
    low_eqbound<-low_eqbound_d*sdpooled
    high_eqbound<-high_eqbound_d*sdpooled
    degree_f<-n1+n2-2
    t1<-(abs(m1-m2)-low_eqbound)/(sdpooled*sqrt(1/n1 + 1/n2))  #students t-test lower bound
    p1<-pt(t1, degree_f, lower=FALSE) 
    t2<-(abs(m1-m2)-high_eqbound)/(sdpooled*sqrt(1/n1 + 1/n2)) #students t-test upper bound
    p2<-pt(t2, degree_f, lower=TRUE) 
    t<-(m1-m2)/(sdpooled*sqrt(1/n1 + 1/n2))
    pttest<-2*pt(-abs(t), df=degree_f)
    LL90<-(m1-m2)-qt(1-alpha, n1+n2-2)*(sdpooled*sqrt(1/n1 + 1/n2))
    UL90<-(m1-m2)+qt(1-alpha, n1+n2-2)*(sdpooled*sqrt(1/n1 + 1/n2))
    LL95<-(m1-m2)-qt(1-(alpha/2), n1+n2-2)*(sdpooled*sqrt(1/n1 + 1/n2))
    UL95<-(m1-m2)+qt(1-(alpha/2), n1+n2-2)*(sdpooled*sqrt(1/n1 + 1/n2))
  } else {
    sdpooled<-sqrt((sd1^2 + sd2^2)/2) #calculate sd root mean squared for Welch's t-test
    low_eqbound<-low_eqbound_d*sdpooled
    high_eqbound<-high_eqbound_d*sdpooled
    degree_f<-(sd1^2/n1+sd2^2/n2)^2/(((sd1^2/n1)^2/(n1-1))+((sd2^2/n2)^2/(n2-1))) #degrees of freedom for Welch's t-test
    t1<-(abs(m1-m2)-low_eqbound)/sqrt(sd1^2/n1 + sd2^2/n2) #welch's t-test upper bound
    p1<-pt(t1, degree_f, lower=FALSE) #p-value for Welch's TOST t-test
    t2<-(abs(m1-m2)-high_eqbound)/sqrt(sd1^2/n1 + sd2^2/n2) #welch's t-test lower bound
    p2<-pt(t2, degree_f, lower=TRUE) #p-value for Welch's TOST t-test
    t<-(m1-m2)/sqrt(sd1^2/n1 + sd2^2/n2) #welch's t-test NHST
    pttest<-2*pt(-abs(t), df=degree_f) #p-value for Welch's t-test
    LL90<-(m1-m2)-qt(1-alpha, degree_f)*sqrt(sd1^2/n1 + sd2^2/n2) #Lower limit for CI Welch's t-test
    UL90<-(m1-m2)+qt(1-alpha, degree_f)*sqrt(sd1^2/n1 + sd2^2/n2) #Upper limit for CI Welch's t-test
    LL95<-(m1-m2)-qt(1-(alpha/2), degree_f)*sqrt(sd1^2/n1 + sd2^2/n2) #Lower limit for CI Welch's t-test
    UL95<-(m1-m2)+qt(1-(alpha/2), degree_f)*sqrt(sd1^2/n1 + sd2^2/n2) #Upper limit for CI Welch's t-test
  }
  ptost<-max(p1,p2) #Get highest p-value for summary TOST result
  ttost<-ifelse(abs(t1) < abs(t2), t1, t2) #Get lowest t-value for summary TOST result
  dif<-(m1-m2)
  testoutcome<-ifelse(pttest<alpha,"significant","non-significant")
  TOSToutcome<-ifelse(ptost<alpha,"significant","non-significant")
  plot(NA, ylim=c(0,1), xlim=c(min(LL90,low_eqbound)-max(UL90-LL90, high_eqbound-low_eqbound)/10, max(UL90,high_eqbound)+max(UL90-LL90, high_eqbound-low_eqbound)/10), bty="l", yaxt="n", ylab="",xlab="Mean Difference")
  points(x=dif, y=0.5, pch=15, cex=2)
  abline(v=high_eqbound, lty=2)
  abline(v=low_eqbound, lty=2)
  abline(v=0, lty=2, col="grey")
  segments(LL90,0.5,UL90,0.5, lwd=3)
  segments(LL95,0.5,UL95,0.5, lwd=1)
  title(main=paste("Equivalence bounds ",round(low_eqbound,digits=3)," and ",round(high_eqbound,digits=3),"\nMean difference = ",round(dif,digits=3)," \n TOST: 90% CI [",round(LL90,digits=3),";",round(UL90,digits=3),"] ", TOSToutcome," \n NHST: 95% CI [",round(LL95,digits=3),";",round(UL95,digits=3),"] ", testoutcome,sep=""), cex.main=1)
  if(var.equal == TRUE) {
    message(cat("Using alpha = ",alpha," Student's t-test was ",testoutcome,", t(",degree_f,") = ",t,", p = ",pttest,sep=""))
    message(cat("Using alpha = ",alpha," the equivalence test based on Student's t-test was ",TOSToutcome,", t(",degree_f,") = ",ttost,", p = ",ptost,sep=""))
  } else {
    message(cat("Using alpha = ",alpha," Welch's t-test was ",testoutcome,", t(",degree_f,") = ",t,", p = ",pttest,sep=""))
    message(cat("Using alpha = ",alpha," the equivalence test based on Welch's t-test  was ",TOSToutcome,", t(",degree_f,") = ",ttost,", p = ",ptost,sep=""))
  }
  TOSTresults<-data.frame(t1,p1,t2,p2,degree_f)
  colnames(TOSTresults) <- c("t-value 1","p-value 1","t-value 2","p-value 2","df")
  bound_d_results<-data.frame(low_eqbound_d,high_eqbound_d)
  colnames(bound_d_results) <- c("low bound d","high bound d")
  bound_results<-data.frame(low_eqbound,high_eqbound)
  colnames(bound_results) <- c("low bound raw","high bound raw")
  CIresults<-data.frame(LL90,UL90)
  colnames(CIresults) <- c(paste("Lower Limit ",100*(1-alpha*2),"% CI raw",sep=""),paste("Upper Limit ",100*(1-alpha*2),"% CI raw",sep=""))
  cat("TOST results:\n")
  print(TOSTresults)  
  cat("\n")
  cat("Equivalence bounds (Cohen's d):\n")
  print(bound_d_results)  
  cat("\n")
  cat("Equivalence bounds (raw scores):\n")
  print(bound_results)  
  cat("\n")
  cat("TOST confidence interval:\n")
  print(CIresults)
  invisible(list(TOST_t1=t1,TOST_p1=p1,TOST_t2=t2,TOST_p2=p2, TOST_df=degree_f,alpha=alpha,low_eqbound=low_eqbound,high_eqbound=high_eqbound,low_eqbound_d=low_eqbound_d,high_eqbound_d=high_eqbound_d, LL_CI_TOST=LL90,UL_CI_TOST=UL90))
}

File: pyTOSTER/tost_two.py
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt

def tost_two(m1, m2, sd1, sd2, n1, n2, low_eqbound_d, high_eqbound_d, alpha=0.05, var_equal=False, plot=True, verbose=True):
    """
    TOST function for an independent t-test (Cohen's d).
    """
    if low_eqbound_d >= high_eqbound_d:
        print("Warning: The lower bound is equal to or larger than the upper bound.")

    if var_equal:
        sdpooled = np.sqrt((((n1 - 1) * (sd1**2)) + ((n2 - 1) * (sd2**2))) / ((n1 + n2) - 2))
    else:
        # sdpooled for Welch's is not used for eqbound calculation in R version
        # sdpooled<-sqrt((sd1^2 + sd2^2)/2) #calculate sd root mean squared for Welch's t-test
        # low_eqbound<-low_eqbound_d*sdpooled
        # high_eqbound<-high_eqbound_d*sdpooled
        # This is not ideal, but to match the R code, we will use the pooled sd for both
        sdpooled = np.sqrt((((n1 - 1) * (sd1**2)) + ((n2 - 1) * (sd2**2))) / ((n1 + n2) - 2))

    low_eqbound = low_eqbound_d * sdpooled
    high_eqbound = high_eqbound_d * sdpooled

    return tost_two_raw(m1, m2, sd1, sd2, n1, n2, low_eqbound, high_eqbound, alpha, var_equal, plot, verbose,
                        low_eqbound_d=low_eqbound_d, high_eqbound_d=high_eqbound_d)


def tost_two_raw(m1, m2, sd1, sd2, n1, n2, low_eqbound, high_eqbound, alpha=0.05, var_equal=False, plot=True, verbose=True, low_eqbound_d=None, high_eqbound_d=None):
    """
    TOST function for an independent t-test (raw scores).
    """
    if low_eqbound >= high_eqbound:
        print("Warning: The lower bound is equal to or larger than the upper bound.")
    if n1 < 2 or n2 < 2:
        raise ValueError("The sample size should be larger than 1.")
    if not (0 < alpha < 1):
        raise ValueError("The alpha level should be a positive value between 0 and 1.")
    if sd1 <= 0 or sd2 <= 0:
        raise ValueError("The standard deviation should be a positive value.")

    if var_equal:
        sdpooled = np.sqrt((((n1 - 1) * (sd1**2)) + ((n2 - 1) * (sd2**2))) / ((n1 + n2) - 2))
        degree_f = n1 + n2 - 2
        se = sdpooled * np.sqrt(1/n1 + 1/n2)
    else:
        degree_f = (sd1**2/n1 + sd2**2/n2)**2 / (((sd1**2/n1)**2/(n1-1)) + ((sd2**2/n2)**2/(n2-1)))
        se = np.sqrt(sd1**2/n1 + sd2**2/n2)

    t = (m1 - m2) / se
    pttest = 2 * (1 - stats.t.cdf(abs(t), df=degree_f))

    t1 = ((m1 - m2) - low_eqbound) / se
    p1 = 1 - stats.t.cdf(t1, df=degree_f)
    t2 = ((m1 - m2) - high_eqbound) / se
    p2 = stats.t.cdf(t2, df=degree_f)
    ttost = t1 if abs(t1) < abs(t2) else t2
    ptost = max(p1, p2)

    diff = m1 - m2
    ll90 = diff - stats.t.ppf(1 - alpha, df=degree_f) * se
    ul90 = diff + stats.t.ppf(1 - alpha, df=degree_f) * se
    ll95 = diff - stats.t.ppf(1 - (alpha / 2), df=degree_f) * se
    ul95 = diff + stats.t.ppf(1 - (alpha / 2), df=degree_f) * se

    test_outcome = "significant" if pttest < alpha else "non-significant"
    tost_outcome = "significant" if ptost < alpha else "non-significant"

    if plot:
        plt.figure()
        plt.ylim(0, 1)
        plot_min = min(ll95, low_eqbound, diff) - max(ul95 - ll95, high_eqbound - low_eqbound, diff) / 10
        plot_max = max(ul95, high_eqbound, diff) + max(ul95 - ll95, high_eqbound - low_eqbound, diff) / 10
        plt.xlim(plot_min, plot_max)

        plt.yticks([])
        plt.xlabel("Mean Difference")
        title = (f"Equivalence bounds {low_eqbound:.3f} and {high_eqbound:.3f}\\n"
                 f"Mean difference = {diff:.3f}\\n"
                 f"TOST: {100*(1-alpha*2):.0f}% CI [{ll90:.3f}; {ul90:.3f}] {tost_outcome}\\n"
                 f"NHST: {100*(1-alpha):.0f}% CI [{ll95:.3f}; {ul95:.3f}] {test_outcome}")
        plt.title(title.replace('\\n', '\n'))

        plt.plot(diff, 0.5, 's', markersize=8, color='black')
        plt.axvline(high_eqbound, linestyle='--', color='black')
        plt.axvline(low_eqbound, linestyle='--', color='black')
        plt.axvline(0, linestyle='--', color='grey')

        plt.hlines(0.5, ll90, ul90, linewidth=3, color='black')
        plt.hlines(0.5, ll95, ul95, linewidth=1, color='black')

        plt.show()

    if verbose:
        print("TOST results:")
        print(f"t-value lower bound: {t1:8.2f}\\tp-value lower bound: {p1:.3f}")
        print(f"t-value upper bound: {t2:8.2f}\\tp-value upper bound: {p2:.3f}")
        print(f"degrees of freedom : {degree_f:.2f}")
        print("\\n")

        if low_eqbound_d is not None and high_eqbound_d is not None:
            print("Equivalence bounds (Cohen's d):")
            print(f"low eqbound: {low_eqbound_d:.4f}\\nhigh eqbound: {high_eqbound_d:.4f}")
            print("\\n")

        print("Equivalence bounds (raw scores):")
        print(f"low eqbound: {low_eqbound:.4f}\\nhigh eqbound: {high_eqbound:.4f}")
        print("\\n")
        print("TOST confidence interval:")
        print(f"lower bound {100*(1-alpha*2):.0f}% CI: {ll90:.3f}\\nupper bound {100*(1-alpha*2):.0f}% CI:  {ul90:.3f}")
        print("\\n")
        print("NHST confidence interval:")
        print(f"lower bound {100*(1-alpha):.0f}% CI: {ll95:.3f}\\nupper bound {100*(1-alpha):.0f}% CI:  {ul95:.3f}")
        print("\\n")
        print("Equivalence Test Result:")
        print(f"The equivalence test was {tost_outcome}, t({degree_f:.2f}) = {ttost:.3f}, p = {ptost:.3f}, given equivalence bounds of {low_eqbound:.3f} and {high_eqbound:.3f} (on a raw scale) and an alpha of {alpha}.")
        print("\\n")
        print("Null Hypothesis Test Result:")
        print(f"The null hypothesis test was {test_outcome}, t({degree_f:.2f}) = {t:.3f}, p = {pttest:.3f}, given an alpha of {alpha}.")

        if pttest <= alpha and ptost <= alpha:
            combined_outcome = ("NHST: reject null significance hypothesis that the effect is equal to 0 \\n"
                                "TOST: reject null equivalence hypothesis")
        elif pttest < alpha and ptost > alpha:
            combined_outcome = ("NHST: reject null significance hypothesis that the effect is equal to 0 \\n"
                                "TOST: don't reject null equivalence hypothesis")
        elif pttest > alpha and ptost <= alpha:
            combined_outcome = ("NHST: don't reject null significance hypothesis that the effect is equal to 0 \\n"
                                "TOST: reject null equivalence hypothesis")
        else: # pttest > alpha and ptost > alpha
            combined_outcome = ("NHST: don't reject null significance hypothesis that the effect is equal to 0 \\n"
                                  "TOST: don't reject null equivalence hypothesis")

        print("\\n")
        print(combined_outcome)

    results = {
        'diff': diff,
        'TOST_t1': t1,
        'TOST_p1': p1,
        'TOST_t2': t2,
        'TOST_p2': p2,
        'TOST_df': degree_f,
        'alpha': alpha,
        'low_eqbound': low_eqbound,
        'high_eqbound': high_eqbound,
        'LL_CI_TOST': ll90,
        'UL_CI_TOST': ul90,
        'LL_CI_TTEST': ll95,
        'UL_CI_TTEST': ul95,
        'NHST_t': t,
        'NHST_p': pttest
    }
    if low_eqbound_d is not None:
        results['low_eqbound_d'] = low_eqbound_d
    if high_eqbound_d is not None:
        results['high_eqbound_d'] = high_eqbound_d

    return results
